src/Temporalio/Worker/WorkflowInstance.cs

```diff
@@ -2230,5 +2266,86 @@ public ExternalWorkflowHandleImpl(WorkflowInstance instance, string id, string?
             public override Task CancelAsync() =>
                 instance.outbound.Value.CancelExternalWorkflowAsync(new(Id: Id, RunId: RunId));
         }
+
+        private class Handlers : LinkedList<Handlers.Handler>
+        {
+#pragma warning disable SA1118 // We're ok w/ string literals spanning lines
+            private static readonly Action<ILogger, string, WarnableSignals, Exception?> SignalWarning =
+                LoggerMessage.Define<string, WarnableSignals>(
+                    LogLevel.Warning,
+                    0,
+                    "Workflow {Id} finished while signal handlers are still running. This may " +
+                    "have interrupted work that the signal handler was doing. You can wait for " +
+                    "all update and signal handlers to complete by using `await " +
+                    "Workflow.WaitConditionAsync(() => Workflow.AllHandlersFinished)`. " +
+                    "Alternatively, if both you and the clients sending the signal are okay with " +
+                    "interrupting running handlers when the workflow finishes, and causing " +
+                    "clients to receive errors, then you can disable this warning via the signal " +
+                    "handler attribute: " +
+                    "`[WorkflowSignal(UnfinishedPolicy=HandlerUnfinishedPolicy.Abandon)]`. The " +
+                    "following signals were unfinished (and warnings were not disabled for their " +
+                    "handler): {Handlers}");
+
+            private static readonly Action<ILogger, string, WarnableUpdates, Exception?> UpdateWarning =
+                LoggerMessage.Define<string, WarnableUpdates>(
```

Comment:
Could maybe combine these two warning functions generically to save on duping the boilerplate part of the message. Not a big deal.